src/web/lib/messagePicker.ts

```diff
@@ -0,0 +1,103 @@
+import { record } from '@root/src/web/browser/ophan/ophan';
+
+export type MaybeFC = React.FC | null;
+type ShowMessage = (meta?: any) => MaybeFC;
+
+export type CanShowResult = {
+	result: boolean;
+	meta?: any;
+};
+
+export type Message = {
+	id: string;
+	show: ShowMessage;
+	canShow: () => Promise<CanShowResult>;
+	timeoutMillis: number | null;
+};
+
+type MessageWithTimeout = Message & {
+	cancelTimeout: () => void;
+};
+
+export type MessageConfig = {
+	messages: Message[];
+	type: string;
+};
+
+const recordMessageTimeoutInOphan = (messageId: string, type: string) =>
+	record({
+		component: `${type}-picker-timeout-dcr`,
+		value: messageId,
+	});
+
+const timeoutify = (message: Message, type: string): MessageWithTimeout => {
+	let timer: number | undefined;
+
+	const canShow = (): Promise<CanShowResult> =>
+		new Promise((resolve) => {
+			if (message.timeoutMillis) {
+				timer = window.setTimeout(() => {
+					recordMessageTimeoutInOphan(message.id, type);
```

Comment:
Similarly here, is it the `message` that times out, or is it the provider? I don't feel strongly here but the other comments have made me don my eagle eyes...